man/RNApdist.1
2.1.7

.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.38.2.
.TH RNAPDIST "1" "April 2014" "RNApdist 2.1.7" "User Commands"
.SH NAME
RNApdist \- manual page for RNApdist 2.1.7
.SH SYNOPSIS
.B RNApdist
[\fIOPTIONS\fR]...
.SH DESCRIPTION
RNApdist 2.1.7
.PP
Calculate distances between thermodynamic RNA secondary structures ensembles
.PP
This program reads RNA sequences from stdin and calculates structure distances
between the thermodynamic ensembles of their secondary structures.
.PP
To do this the partition function and matrix of base pairing probabilities is
computed for each sequence. The probability matrix is then condensed into a
vector holding for each base the probabilities of being unpaired, paired
upstream, or paired downstream, respectively. These profiles are compared
by a standard alignment algorithm.

The base pair probabilities are also saved as postscript "dot plots" (as in
RNAfold) in the files  "name_dp.ps", where name is the name of the sequence,
or a number if unnamed.
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help and exit
.TP
\fB\-\-detailed\-help\fR
Print help, including all details and hidden
options, and exit
.TP
\fB\-\-full\-help\fR
Print help, including hidden options, and exit
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version and exit
.SS "General Options:"
.TP
\fB\-\-noconv\fR
Do not automatically substitude nucleotide
"T" with "U"
.IP
(default=off)
.SS "Algorithms:"
.TP
\fB\-X\fR, \fB\-\-compare\fR=\fIp\fR|m|f|c
Specify the comparison directive.
(default=`p')
.IP
Possible arguments for this option are: \fB\-Xp\fR compare the structures pairwise
(p), i.e. first with 2nd, third with 4th etc.
\fB\-Xm\fR calculate the distance matrix between all structures. The output is
formatted as a lower triangle matrix.
\fB\-Xf\fR compare each structure to the first one.
\fB\-Xc\fR compare continuously, that is i\-th with (i+1)th structure.
.TP
\fB\-B\fR, \fB\-\-backtrack[=\fR<filename>]
Print an "alignment" with gaps of the
profiles. The aligned structures are written
to <filename>, if specified.
.IP
(default=`none')
.IP
Within the profile output, the following symbols will be used:
.TP
()
essentially upstream (downstream) paired bases
.TP
{}
weakly upstream (downstream) paired bases
.TP
|
strongly paired bases without preference
.TP
,
weakly paired bases without preference
.TP
\&.
essentially unpaired bases.
.IP
If <filename> is not specified, the output is written to stdout, unless the
.IP
"\-Xm" option is set in which case "backtrack.file" is used.
.SS "Model Details:"
.TP
\fB\-T\fR, \fB\-\-temp\fR=\fIDOUBLE\fR
Rescale energy parameters to a temperature of
temp C. Default is 37C.
.TP
\fB\-4\fR, \fB\-\-noTetra\fR
Do not include special tabulated stabilizing
energies for tri\-, tetra\- and hexaloop
hairpins. Mostly for testing.
.IP
(default=off)
.TP
\fB\-d\fR, \fB\-\-dangles\fR=\fIINT\fR
set energy model for treatment of dangling
bases
.IP
(possible values="0", "2" default=`2')
.TP
\fB\-\-noLP\fR
Produce structures without lonely pairs
(helices of length 1).
.IP
(default=off)
.IP
For partition function folding this only disallows pairs that can only occur
isolated. Other pairs may still occasionally occur as helices of length 1.
.TP
\fB\-\-noGU\fR
Do not allow GU pairs
.IP
(default=off)
.TP
\fB\-\-noClosingGU\fR
Do not allow GU pairs at the end of helices
.IP
(default=off)
.TP
\fB\-P\fR, \fB\-\-paramFile\fR=\fIparamfile\fR
Read energy parameters from paramfile, instead
of using the default parameter set.
.IP
A sample parameter file should accompany your distribution.
See the RNAlib documentation for details on the file format.
.TP
\fB\-\-nsp\fR=\fISTRING\fR
Allow other pairs in addition to the usual
AU,GC,and GU pairs.
.IP
Its argument is a comma separated list of additionally allowed pairs. If the
first character is a "\-" then AB will imply that AB and BA are allowed
pairs.
e.g. RNAfold \fB\-nsp\fR \fB\-GA\fR  will allow GA and AG pairs. Nonstandard pairs are
given 0 stacking energy.
.TP
\fB\-e\fR, \fB\-\-energyModel\fR=\fIINT\fR
Rarely used option to fold sequences from the
artificial ABCD... alphabet, where A pairs B,
C\-D etc.  Use the energy parameters for GC
(\fB\-e\fR 1) or AU (\fB\-e\fR 2) pairs.
.SH AUTHOR

Peter F Stadler, Ivo L Hofacker, Sebastian Bonhoeffer.
.SH REFERENCES
.I If you use this program in your work you might want to cite:

R. Lorenz, S.H. Bernhart, C. Hoener zu Siederdissen, H. Tafer, C. Flamm, P.F. Stadler and I.L. Hofacker (2011),
"ViennaRNA Package 2.0",
Algorithms for Molecular Biology: 6:26 

I.L. Hofacker, W. Fontana, P.F. Stadler, S. Bonhoeffer, M. Tacker, P. Schuster (1994),
"Fast Folding and Comparison of RNA Secondary Structures",
Monatshefte f. Chemie: 125, pp 167-188


S. Bonhoeffer, J.S. McCaskill, P.F. Stadler, P. Schuster (1993),
"RNA multi-structure landscapes",
Euro Biophys J:22, pp 13-24

.I The energy parameters are taken from:

D.H. Mathews, M.D. Disney, D. Matthew, J.L. Childs, S.J. Schroeder, J. Susan, M. Zuker, D.H. Turner (2004),
"Incorporating chemical modification constraints into a dynamic programming algorithm for prediction of RNA secondary structure",
Proc. Natl. Acad. Sci. USA: 101, pp 7287-7292

D.H Turner, D.H. Mathews (2009),
"NNDB: The nearest neighbor parameter database for predicting stability of nucleic acid secondary structure",
Nucleic Acids Research: 38, pp 280-282
.SH "REPORTING BUGS"
If in doubt our program is right, nature is at fault.
.br
Comments should be sent to rna@tbi.univie.ac.at.


NEWS
Some Changes relative to ViennaRNA-1.03

New energy parameters used in the folding routines.
Prediction of minimum free energy structures now includes contributions
from terminal unpaired nucleotides (so called dangling ends).
The partition function treats these contributions approximately.

Energies for GU pairs in are now taken from
L. He et al., Biochemistry 1991, 30, p 11124

Bulges of length 1 now get an additional stacking energy.

The folding routines accept  a list of nonstandard basepairs that should be
allowed to occur. Also, when doing constrained folding, forced pairs will
be honoured even if the pair is usally not allowed (not AU, GC, or GU).
RNAeval should be useful now for structures containing nonstandard pairs.

Some support for constrained folding using partition function: A constraint
of the form i-j must pair or i must not pair will be honoured by disallowing 
all pairs conflicting with the constraint.

Both folding algorthms use less memory and should be a bit faster for long
sequences.

Several smaller bug fixes.


Progs/RNApdist.c
/*
		Distances of Secondary Structure Ensembles
	  Peter F Stadler, Ivo L Hofacker, Sebastian Bonhoeffer
			Vienna RNA Package
*/
#include <stdio.h>
#include <math.h>
#include <ctype.h>
#include <malloc.h>
#include <errno.h>
#include <string.h>
#include "part_func.h"
#include "fold_vars.h"
#include "profiledist.h"
#include "dist_vars.h"
#include "utils.h"

#define PUBLIC
#define PRIVATE    static

#define MAXLENGTH  10000
#define MAXSEQ      1000

static char rcsid[] = "$Id: RNApdist.c,v 1.3 1997/11/06 17:40:46 ivo Exp $";

PRIVATE void command_line(int argc, char *argv[]);
PRIVATE void usage(void);
PRIVATE void print_aligned_lines(FILE *somewhere);

PRIVATE char task;
PRIVATE char  ParamFile[256]="";
PRIVATE char outfile[50];
PRIVATE char  ruler[] ="....,....1....,....2....,....3....,....4"
                       "....,....5....,....6....,....7....,....8";

extern void PS_dot_plot(char *string, char *file);

int main(int argc, char *argv[])
     
{
   float   **T[MAXSEQ];
   int        i,j, istty, n=0;
   int        type, length, taxa_list=0;
   float      dist;
   FILE      *somewhere;
   char      *structure;
   char      *line=NULL, *cp, fname[20], *list_title=NULL;
   
   command_line(argc, argv);
   
   if((outfile[0]=='\0')&&(task=='m')&&(edit_backtrack)) 
      strcpy(outfile,"backtrack.file"); 
   if(outfile[0]!='\0') somewhere = fopen(outfile,"w");
   else somewhere = stdout;
   istty   = (isatty(fileno(stdout))&&isatty(fileno(stdin)));

   if (ParamFile[0])
     read_parameter_file(ParamFile);

   while (1) {
      if ((istty)&&(n==0)) {
	 printf("\nInput sequence;  @ to quit\n");
	 printf("%s\n", ruler);
      }

      type = 0;
      do {  /* get sequence to fold */
	 if (line!=NULL) free(line);
	 if ((line=get_line(stdin))==NULL) {type = 999; break;}
         if (line[0]=='@') type = 999;
	 if (line[0]=='*') {
	    if (taxa_list==0) {
	       if (task=='m') taxa_list=1;
	       printf("%s\n", line);
	       type = 0;
	    } else {
	       list_title = strdup(line);
	       type = 888;
	    }
	 }
         if (line[0]=='>') {
	    *fname='\0';
	    sscanf(line,">%12s", fname);
	    if (*fname) strcat(fname, "_dp.ps");
	    if (taxa_list)
	       printf("%d : %s\n", n+1, line+1);
	    else printf("%s\n",line);
	    type = 0;
	 }
         if (isalpha(line[0]))  {
	    cp =strchr(line,' ');
	    if (cp) *cp='\0';
	    type = 1;
         }
      } while(type==0);

      if( (task == 'm')&&(type>800) ) {
	 if (taxa_list) 
	    printf("* END of taxa list\n");
         printf("> p %d (pdist)\n",n);
         for (i=1; i<n; i++) {
            for (j=0; j<i; j++) {
               printf("%g ",profile_edit_distance(T[i], T[j]));
               if(edit_backtrack) fprintf(somewhere,"> %d %d\n",i+1,j+1);
               print_aligned_lines(somewhere);
	    }
	    printf("\n");
	 }
	 if (type==888) {  /* do another distance matrix */
	    n = 0;
	    printf("%s\n", list_title);
	    free(list_title);
	 }
      }      
      
      if(type>800) {
	 for (i=0; i<n; i++) 
	    free_profile(T[i]);
	 if (type == 888) continue;
	 if (outfile[0]!='\0') fclose(somewhere); 
         return 0; /* finito */
      }
      
      length = strlen(line);
      for (i=0; i<length; i++) line[i]=toupper(line[i]);
      
      init_pf_fold(length);
      structure = (char *) space((length+1)*sizeof(char));
      pf_fold(line,structure);

      if (*fname=='\0')
	 sprintf(fname, "%d_dp.ps", n+1);
      PS_dot_plot(line, fname);
      *fname='\0';

      T[n] = Make_bp_profile(length);
      if((istty)&&(task=='m')) printf("%s\n",structure);
      free(structure);
      
      free_pf_arrays();
      
      n++;
      switch (task) {
       case 'p' :
	  if (n==2) {
	     dist = profile_edit_distance(T[0],T[1]);
	     printf("%g\n",dist);
	     print_aligned_lines(somewhere);
	     free_profile(T[0]);
	     free_profile(T[1]);
	     n=0;
	  }
	  break;
	case 'f' :
	   if (n>1) { 
	      dist = profile_edit_distance(T[1], T[0]);
	      printf("%g\n",dist);
	      print_aligned_lines(somewhere);
	      free_profile(T[1]);
	      n=1;
	   }
	  break;
	case 'c' :
	   if (n>1) {
	      dist = profile_edit_distance(T[1], T[0]);
	      printf("%g\n",dist);
	      print_aligned_lines(somewhere);
	      free_profile(T[0]);
	      T[0] = T[1]; 
	      n=1; 
	   }
	  break;
	  
	case 'm' : 
	   break;
	  
	  default :
	     nrerror("This can't happen.");
       }    /* END switch task */
      fflush(stdout);
   }    /* END while */
}

/* ----------------------------------------------------------------- */

PRIVATE void command_line(int argc, char *argv[])
{

   int i, sym, r;
   char  ns_bases[33]="", *c;

   task = 'p';
   for (i=1; i<argc; i++) {
      if (argv[i][0]=='-') {
	 switch (argv[i][1]) {
	  case 'T':      /* temperature for folding */
             if (argv[i][2]!='\0') usage();
	     if (sscanf(argv[++i], "%f", &temperature)==0)
	       usage();
	     break;
	  case '4':
	     tetra_loop=0;
	     break;
 	  case 'd':
	     dangles=0;
	     break;
	  case 'e':
	     if (sscanf(argv[++i],"%d", &energy_set)==0)
	       usage();
	     break;
	  case 'n':
            if ( strcmp(argv[i], "-noGU" )==0) noGU=1;
            if ( strcmp(argv[i], "-noCloseGU" ) ==0) no_closingGU=1;
            if ( strcmp(argv[i], "-nsp") ==0) {
	      if (i==argc-1) usage();
              r=sscanf(argv[++i], "%32s", ns_bases);
              if (!r) usage();
            }
            break;
	  case 'X':
	    switch (task = argv[i][2]) {
	     case 'p': break;
	     case 'm': break;
	     case 'f': break;
	     case 'c': break;
             default : usage();
	    }
	    break;
          case 'B':
            if(argv[i][2]!='\0') usage();
            if( (i+1) >= argc) outfile[0] = '\0';
            else if (argv[i+1][0]=='-') outfile[0] = '\0';
            else {
               i++;
               strncpy(outfile,argv[i],49);
            }
            edit_backtrack = 1;   
	    break;
	  case 'P':
	    if (sscanf(argv[++i], "%255s", ParamFile)==0)
	      usage();
	    break;
	  default:
	    usage();
	 }
      }
   }
   if (ns_bases[0]) {
      nonstandards = space(33);
      c=ns_bases;
      i=sym=0;
      if (*c=='-') {
         sym=1; c++;
      }
      while (*c) {
         if (*c!=',') {
            nonstandards[i++]=*c++;
            nonstandards[i++]=*c;
            if ((sym)&&(*c!=*(c-1))) {
               nonstandards[i++]=*c;
               nonstandards[i++]=*(c-1);
            }
         }
         c++;
      }
   }
}

/* ---------------------------------------------------------------------------- */

PRIVATE void usage(void)
{
  nrerror("usage: RNApdist [-Xpmfc] [-B [file]] [-T temp] [-4] [-d] [-noGU]\n"
	  "                [-noCloseGU] [-e e_set] [-P paramfile] [-nsp pairs]");
}

/*--------------------------------------------------------------------------*/

PRIVATE void print_aligned_lines(FILE *somewhere)
{
   if (edit_backtrack)
     fprintf(somewhere, "%s\n%s\n", aligned_line[0], aligned_line[1]);
}

/*--------------------------------------------------------------------------*/
